### PR TITLE
CORENET-6130, CORENET-6261, CORENET-6092: Implement PreconfiguredUDNAddresses API changes

### PR DESCRIPTION
**Pull request by kyrtapz**
Introduce the following changes behind the featuregate:
- Layer2 (Cluster) UDN API update: Add new API fields to both CUDN and UDN CRDs behind the PreconfiguredUDNAddresses featuregate
- Add ValidatingAdmissionPolicy blocking v1.multus-cni.io/default-network updates: Prevent modification of v1.multus-cni.io/default-network once a pod is created, applies to environments with PreconfiguredUDNAddresses featuregate enabled
-  Avoid multus-admission webhook race with ovn-kubernetes: Add CEL expression to ignore default/openshift-ovn-kubernetes NAD to prevent circular dependency where ovn-k fails to start because multus webhook blocks NAD creation, while webhook uses cluster-networked pods which require ovn-k to be running

### CODE REVIEW

**Comment by maiqueb** [comment]
@kyrtapz could you rebase, now that #2770 was merged ? 

I see these warnings being logged in the ovnkube control plane pods:
```
2025-08-11T14:14:07.099359183Z E0811 14:14:07.099115       1 obj_retry.go:681] Failed to update *v1.Pod, old=e2e-network-segmentation-e2e-3110/virt-launcher-myvm-wm4dt, new=e2e-network-segmentation-e2e-3110/virt-launcher-myvm-wm4dt, error: failed to update pod e2e-network-segmentation-e2e-3110/virt-launcher-myvm-wm4dt: error retrieving IPAMClaim for pod e2e-network-segmentation-e2e-3110/virt-launcher-myvm-wm4dt: failed to get IPAMClaim "myvm.overlay": ipamclaim.k8s.cni.cncf.io "myvm.overlay" not found
2025-08-11T14:14:07.139482397Z W0811 14:14:07.137061       1 warnings.go:70] unknown field "status.ownerPod"
```

I think it's a red herring, but still, it would give us a cleaner signal. 

Thanks in advance

**Comment by kyrtapz** [comment]
`verify` failed due to lint issues, will re-push once we get some signal from the techpreview job.

**Comment by kyrtapz** [comment]
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2743/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview/1958143914649063424/artifacts/e2e-gcp-ovn-techpreview/gather-extra/artifacts/pods/openshift-network-operator_network-operator-57d5656d84-6zjq6_network-operator.log
CNO was good in this job: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2743/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview/1958143914649063424/

/test e2e-gcp-ovn-techpreview

**Comment by kyrtapz** [comment]
Fixed a corner case in the VAP to handle pods that do not have annotations field at all: https://github.com/openshift/cluster-network-operator/compare/59a03f2d2ba6fb823123671e774d938f12c77d7a..79eea1b035a472a9b27a9be85460e730fe94c2b3

File: bindata/network/ovn-kubernetes/common/001-crd.yaml [hunk 1]
```diff
@@ -3283,6 +3283,50 @@ spec:
               layer2:
                 description: Layer2 is the Layer2 topology configuration.
                 properties:
+{{- if .OVN_PRE_CONF_UDN_ADDR_ENABLE }}
+                  defaultGatewayIPs:
```
Comment (tssurya) [inline bindata/network/ovn-kubernetes/common/001-crd.yaml:3287]: this looks exact same to me from https://github.com/ovn-kubernetes/ovn-kubernetes/blob/7c047281d46a6e5f154a0f5d3feb8866af3619f4/dist/templates/k8s.ovn.org_clusteruserdefinednetworks.yaml.j2#L97

File: bindata/network/ovn-kubernetes/common/default-net-annotation-policy.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        apiVersions: ["v1"]
+        operations:  ["UPDATE"]
+        resources:   ["pods"]
+  failurePolicy: Fail
+  validations:
+    # Prevent any changes to the default-network annotation after pod creation:
+    # - If annotation exists in old pod: new pod must have same annotation with identical value
+    # - If annotation doesn't exist in old pod: new pod must also not have it
+    - expression: >
+        !has(object.metadata.annotations) || !has(oldObject.metadata.annotations) ||
```
Comment (tssurya) [inline bindata/network/ovn-kubernetes/common/default-net-annotation-policy.yaml:19]: slight change here that is missing u/s as noted by @kyrtapz already in his comment on the PR. This fix will also go upstream.

File: pkg/network/ovn_kubernetes_test.go [hunk 3]
```diff
@@ -4187,11 +4187,11 @@ func Test_renderOVNKubernetes(t *testing.T) {
...
 				}, bootstrapResult: fakeBootstrapResultOVN,
-				manifestDir:  manifestDirOvn,
-				client:       cnofake.NewFakeClient(),
-				featureGates: preDefUDNFeatureGates,
+				manifestDir:        manifestDirOvn,
+				client:             cnofake.NewFakeClient(),
+				featureGates:       preDefUDNFeatureGates,
 			},
-			expectNumObjs: 45,
+			expectNumObjs: 47,
```
Comment (tssurya) [inline pkg/network/ovn_kubernetes_test.go:4194]: ah the VAP and the role binding?

File: bindata/network/multus-admission-controller/003-webhook.yaml [hunk 4]
```diff
@@ -31,6 +31,11 @@ webhooks:
       # On updates, only validate if the Spec changes
       - name: CreateDeleteOrUpdatedSpec
         expression: oldObject == null || object == null || has(object.spec) != has(oldObject.spec) || (has(object.spec) && object.spec != oldObject.spec)
+{{if .OVN_PRE_CONF_UDN_ADDR_ENABLE}}
+      # Ignore default/openshift-ovn-kubernetes NAD to avoid a race between ovn-kubernetes and the multus webhook on install
+      - name: IgnoreDefaultOVNKubernetesNAD
+        expression: object == null || object.metadata.namespace != "openshift-ovn-kubernetes" || object.metadata.name != "default"
```
Comment (tssurya) [inline bindata/network/multus-admission-controller/003-webhook.yaml:37]: haha in my head i go:

ignore if:
object != null && object.metadata.namespace == "openshift-ovn-kubernetes" && object.metadata.name == "default"

but i see you did exact opp which is the same thing

Comment (tssurya) [reply bindata/network/multus-admission-controller/003-webhook.yaml:37]: i really hope we can't have a NAD with no namespace and metadata.namespace is never none haha like we hit the hitch with metadata.annotations

**Review by tssurya — APPROVED** [review]
/lgtm

changes look good to me. not sure if you want anyone from multus side to ack it but should be ok

**Comment by tssurya** [comment]
/hold

@kyrtapz feel free to remove this whenever you are happy with CI and testing

File: bindata/network/multus-admission-controller/003-webhook.yaml (hunk 4 above)
Comment (kyrtapz) [reply bindata/network/multus-admission-controller/003-webhook.yaml:37]: Great point, I will double check that this doesn't happen.

**Comment by kyrtapz** [comment]
```
: [sig-auth][Feature:ProjectAPI] TestProjectWatch should succeed [apigroup:project.openshift.io][apigroup:authorization.openshift.io][apigroup:user.openshift.io] [Suite:openshift/conformance/parallel] expand_less
Run #0: Failed expand_less 	5m24s
{  fail [github.com/openshift/origin/test/extended/project/project.go:239]: timeout: e2e-test-project-api-8lmdk}
[open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20250821-102018.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20250821-102018.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%2C%22prowjob_junit.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Fpr-logs%2Fpull%2Fopenshift_cluster-network-operator%2F2743%2Fpull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview%2F1958457821871738880%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2743/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview/1958457821871738880&lensIndex=2#)
Run #1: Failed expand_less 	5m10s
{  fail [github.com/openshift/origin/test/extended/project/project.go:239]: timeout: e2e-test-project-api-4kggm}
[open stdoutopen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fgather-must-gather%2Fartifacts%2Fjunit_install.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fe2e-monitor-tests__20250821-102018.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit%2Fjunit_e2e__20250821-102018.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_node_ready.xml%22%2C%22artifacts%2Fe2e-gcp-ovn-techpreview%2Fopenshift-e2e-test%2Fartifacts%2Fjunit_nodes.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%2C%22prowjob_junit.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Ftest-platform-results%2Fpr-logs%2Fpull%2Fopenshift_cluster-network-operator%2F2743%2Fpull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview%2F1958457821871738880%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2743/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview/1958457821871738880&lensIndex=2#)
```
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2743/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-ovn-techpreview/1958457821871738880
Very unlikely that this PR caused it, retesting.

/test e2e-gcp-ovn-techpreview

**Comment by kyrtapz** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.20-e2e-aws-ovn-techpreview

**Comment by kyrtapz** [comment]
aws-techpreview is passing: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-cluster-network-operator-2743-ci-4.20-e2e-aws-ovn-techpreview/1958542926258638848
/hold cancel
/test e2e-gcp-ovn-techpreview

**Comment by tssurya** [comment]
/override e2e-metal-ipi-ovn-dualstack-bgp-local-gw
because of https://issues.redhat.com/browse/OCPBUGS-60455

**Comment by tssurya** [comment]
/override ci/prow/e2e-metal-ipi-ovn-dualstack-bgp-local-gw
because of https://issues.redhat.com/browse/OCPBUGS-60455

**Comment by kyrtapz** [comment]
Changes in this PR are TechPreview specific and should not affect this job, will re-run it anyway once tide kicks of the other 20 jobs... 
/override ci/prow/e2e-metal-ipi-ovn-ipv6-ipsec

**Comment by kyrtapz** [comment]
Metal jobs are timing out, this change is techpreview specific and not affecting ipsec.

/override ci/prow/e2e-metal-ipi-ovn-ipv6-ipsec

**Comment by kyrtapz** [comment]
/override ci/prow/e2e-metal-ipi-ovn-dualstack-bgp-local-gw
because of https://issues.redhat.com/browse/OCPBUGS-60455

**Comment by kyrtapz** [comment]
This PR only affects techpreview, metal jobs are perma-failing (timeout) across the board:
/override ci/prow/e2e-metal-ipi-ovn-dualstack-bgp
/override ci/prow/e2e-metal-ipi-ovn-dualstack-bgp-local-gw
/override ci/prow/e2e-metal-ipi-ovn-ipv6

gcp techpreview failed on CSI test which is unrelated:
/override ci/prow/e2e-gcp-ovn-techpreview
